lxd/container_console.go

```diff
@@ -0,0 +1,483 @@
+	if !c.IsRunning() {
+		return BadRequest(err)
+	}
+
+	err = fmt.Errorf("Container is frozen")
+	if c.IsFrozen() {
+		return BadRequest(err)
+	}
+
+	post := api.ContainerConsolePost{}
+	buf, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	err = json.Unmarshal(buf, &post)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	ws := &consoleWs{}
+	ws.fds = map[int]string{}
+
+	idmapset, err := c.IdmapSet()
+	if err != nil {
+		return InternalError(err)
+	}
+
+	if idmapset != nil {
+		ws.rootUid, ws.rootGid = idmapset.ShiftIntoNs(0, 0)
+	}
+
+	ws.conns = map[int]*websocket.Conn{}
+	ws.conns[-1] = nil
+	ws.conns[0] = nil
+	for i := -1; i < len(ws.conns)-1; i++ {
+		ws.fds[i], err = shared.RandomCryptoString()
+		if err != nil {
+			return InternalError(err)
+		}
+	}
+
+	ws.allConnected = make(chan bool, 1)
+	ws.controlConnected = make(chan bool, 1)
+
+	ws.container = c
+	ws.width = post.Width
+	ws.height = post.Height
+
+	resources := map[string][]string{}
+	resources["containers"] = []string{ws.container.Name()}
+
+	op, err := operationCreate(operationClassWebsocket, resources,
+		ws.Metadata(), ws.Do, nil, ws.Connect)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	return OperationResponse(op)
+}
+
+func containerConsoleLogGet(d *Daemon, r *http.Request) Response {
+	name := mux.Vars(r)["name"]
+	c, err := containerLoadByName(d.State(), name)
+	if err != nil {
+		return SmartError(err)
+	}
+
+	expandedConfig := c.ExpandedConfig()
+	rawLxc := expandedConfig["raw.lxc"]
+
+	consoleLogpath := ""
+	for _, line := range strings.Split(rawLxc, "\n") {
+		key, val, err := lxcParseRawLXC(line)
+		if err != nil {
+			return SmartError(err)
+		}
+
+		if key != "lxc.console.logfile" {
+			continue
+		}
+
+		consoleLogpath = val
+		break
+	}
+
+	logContents := ""
+	console := lxc.ConsoleLogOptions{}
+
+	ent := fileResponseEntry{}
+	if !c.IsRunning() {
+		if consoleLogpath == "" {
+			return SmartError(fmt.Errorf("The container does not keep a console log"))
+		}
+
+		// Hand back the contents of the on-disk logfile.
+		ent.path = consoleLogpath
+		ent.filename = consoleLogpath
+		goto onSuccess
+	}
+
+	// Container keeps an on-disk logfile so keep sync it here.
+	if consoleLogpath != "" {
+		console.WriteToLogFile = true
+	} else {
+		console.WriteToLogFile = false
+	}
+
+	// Query the container's console ringbuffer.
+	console.ClearLog = false
+	console.ReadLog = true
+	console.ReadMax = 0
+
+	// Send a ringbuffer request to the container.
+	logContents, err = c.ConsoleLog(console)
+	if err != nil {
+		errno, isErrno := shared.GetErrno(err)
+		if !isErrno {
+			return SmartError(err)
+		}
+
+		if errno == syscall.ENODATA {
+			goto onSuccess
+		}
+
+		if errno == syscall.EFAULT && consoleLogpath != "" {
+			// Hand back the contents of the on-disk logfile.
+			ent.path = consoleLogpath
+			ent.filename = consoleLogpath
+			goto onSuccess
+		}
+
+		return SmartError(err)
+	}
+
+	ent.buffer = []byte(logContents)
+
+onSuccess:
```

Comment:
Not sure how I feel about the use of gotos in Go... looks like you've already used them in a few spots in the storage codebase (which I must have missed during review).

In general I try to stay away from goto as it makes the code very hard to read, especially when you factor in Go's already annoying variable scoping problem.

I've found it reasonable in the LXC C codebase and even in LXD's cgo codebase for memory management and error handling but Go doesn't really come with that same flow and so I think they're usually unneeded and the benefit they may provide doesn't make up for the added difficulty in reading the code.


I'd recommend we stay away from them in the LXD Go codebase and remove the few ones we already have today.

Reply:
done (apart from removal across the code base. could be a nice starter issue for someone)